File: protocol-generator/dart/lib/src/conformance/harness.dart
import 'dart:io';
import 'dart:typed_data';

import '../codec/ecf.dart';
import '../codec/ecf_value.dart';
import '../crypto/content_hash.dart';
import '../crypto/ed.dart';
import '../crypto/peer_id.dart';
import '../errors.dart';

/// ECF wire-conformance harness (the codec gate).
///
/// The normative fixture `conformance-vectors-v1.cbor` is itself a canonical-ECF
/// array of vector maps, each carrying its own cross-blessed `canonical` bytes
/// (the Go `wire-conformance` oracle output, 3-way Go × Rust × Python
/// byte-locked). The harness decodes the fixture with THIS peer's OWN decoder (a
/// decoder bug is itself a conformance failure per ENTITY-CBOR-ENCODING.md §E.3),
/// runs each vector through the codec, and byte-compares against the embedded
/// `canonical`. Byte-identity to the fixture == oracle PASS.
///
/// Dispatch by `kind` + `id` category prefix:
///  - decode_reject -> the decoder MUST reject the canonical wire bytes
///  - encode_equal, category:
///      content_hash -> varint(format_code) ‖ SHA-2(ECF({type,data}))
///      peer_id      -> CBOR-text(Base58(varint(kt)‖varint(ht)‖digest))
///      signature    -> Ed25519_sign(seed, ECF({type,data}))  [raw 64 bytes]
///      else         -> plain ECF encode(input)
class ConformanceHarness {
  /// Resolve the fixture path: env `ECF_FIXTURE` or the vendored default.
  static String defaultFixture() {
    final env = Platform.environment['ECF_FIXTURE'];
    if (env != null && env.isNotEmpty) return env;
    return '../shared/test-vectors/v0.8.0/conformance-vectors-v1.cbor';
  }

  /// Run the corpus at [fixturePath]. Returns a [ConformanceResult].
  static Future<ConformanceResult> run(String fixturePath) async {
    final octets = await File(fixturePath).readAsBytes();
    final decoded = Ecf.decode(Uint8List.fromList(octets));
    if (decoded is! Ok<EcfValue>) {
      throw StateError(
          'fixture failed to decode with our own decoder: ${(decoded as Err).error}');
    }
    final top = decoded.value;
    if (top is! EcfArray) {
      throw StateError('fixture top-level is not an array');
    }

    var pass = 0;
    var fail = 0;
    var total = 0;
    final failures = <String>[];

    for (final v in top.items) {
      if (v is! EcfMap) continue; // meta / non-vector
      final kind = _text(v['kind']);
      if (kind == null) continue; // meta entry without a kind
      final id = _text(v['id']) ?? '?';
      total++;
      bool ok;
      String? detail;
      try {
        switch (kind) {
          case 'decode_reject':
            final wire = _bytes(v['canonical']);
            ok = Ecf.decode(wire) is Err;
            if (!ok) detail = 'decoder ACCEPTED a reject vector';
          case 'encode_equal':
            final want = _bytes(v['canonical']);
            final got = await _produce(id, v['input']!);
            ok = _bytesEqual(got, want);
            if (!ok) detail = 'want=${_hex(want)} got=${_hex(got)}';
          default:
            total--;
            continue;
        }
      } catch (e) {
        ok = false;
        detail = 'raised: $e';
      }
      if (ok) {
        pass++;
      } else {
        fail++;
        failures.add('FAIL $id: $detail');
      }
    }
    return ConformanceResult(pass, fail, total, failures);
  }

  static Future<Uint8List> _produce(String id, EcfValue input) async {
    switch (_category(id)) {
      case 'content_hash':
        final m = input as EcfMap;
        final codeVal = m['format_code'];
        final code = codeVal is EcfInt ? codeVal.value.toInt() : 0;
        final entity = EcfMap.of({'type': m['type']!, 'data': m['data']!});
        return ContentHash.compute(entity, formatCode: code);
      case 'peer_id':
        final m = input as EcfMap;
        final kt = (m['key_type']! as EcfInt).value.toInt();
        final ht = (m['hash_type']! as EcfInt).value.toInt();
        final digest = _bytes(m['digest']);
        final peerId = PeerId.format(kt, ht, digest);
        // canonical = the peer_id string encoded as a CBOR text string
        return Ecf.encodeOrThrow(EcfText(peerId));
      case 'signature':
        final m = input as EcfMap;
        final seed = _bytes(m['seed']);
        final entity = m['entity']! as EcfMap;
        final hashed =
            EcfMap.of({'type': entity['type']!, 'data': entity['data']!});
        final ecf = Ecf.encodeOrThrow(hashed);
        return Ed.sign(seed, ecf, Curve.ed25519);
      default:
        return Ecf.encodeOrThrow(input);
    }
  }

  static String _category(String id) {
    final dot = id.indexOf('.');
    return dot >= 0 ? id.substring(0, dot) : id;
  }

  static String? _text(EcfValue? v) => v is EcfText ? v.value : null;

  static Uint8List _bytes(EcfValue? v) {
    if (v is EcfBytes) return v.octets;
    throw StateError('expected bytes, got $v');
  }

  static bool _bytesEqual(Uint8List a, Uint8List b) {
    if (a.length != b.length) return false;
    for (var i = 0; i < a.length; i++) {
      if (a[i] != b[i]) return false;
    }
    return true;
  }

  static String _hex(Uint8List b) =>
      b.map((x) => x.toRadixString(16).padLeft(2, '0')).join();
}

/// The outcome of a conformance run.
class ConformanceResult {
  const ConformanceResult(this.pass, this.fail, this.total, this.failures);
  final int pass;
  final int fail;
  final int total;
  final List<String> failures;
}

File: protocol-generator/python/tests/conformance/harness.py
"""ECF wire-conformance harness — the S2 conformance gate for the Python peer.

Loads the normative cross-blessed corpus
(``protocol-generator/shared/test-vectors/v0.8.0/conformance-vectors-v1.cbor``)
through the peer's OWN ECF decoder (a decoder bug surfaces here first, per
Appendix E §E.3 step 1), then for each vector branches on ``kind``:

  * ``encode_equal`` — run ``input`` through the peer's construction for the
    vector's category and assert the bytes are byte-identical to the vector's
    ``canonical`` field.
  * ``decode_reject`` — feed ``canonical`` (non-canonical wire bytes) to the
    peer's strict ECF decoder; pass iff it rejects.

Category dispatch mirrors the Go oracle (``entity-core-go`` @ 33f35fd,
``cmd/internal/wire-conformance/emit.go`` :: ``encodeVector``) EXACTLY — this
is a clean-room BYTE-validation against the oracle's construction (we read its
construction contract, not to copy code, but because the cross-blessed
``canonical`` bytes ARE the contract):

  * ``content_hash.*`` — ``varint(format_code) || SHA256(ECF({type, data}))``
    (``format_code`` defaults to 0; an override key exercises multi-byte varint)
  * ``peer_id.*``      — ``ECF(Base58(varint(kt) || varint(ht) || digest))``
    (the canonical bytes are the ECF text-string of the Base58 string)
  * ``signature.*``    — Ed25519 sign over ``ECF(entity)`` (deterministic seed)
  * ``envelope.*``     — canonical ECF re-encode of the envelope value tree
  * everything else    — Class A canonical ECF re-encode of ``input``

Run as a script (``python -m tests.conformance.harness [corpus.cbor]``) to
get a JSON+text report on stdout; importable for the pytest suite.
"""

from __future__ import annotations

import hashlib
import json
import sys
from pathlib import Path
from typing import Any

from entity_core import _base58, _cbor
from entity_core._cbor import ByteKey
from entity_core._varint import encode_varint
from entity_core.signature import sign_ed25519

CORPUS_VERSION = "v1"
SPEC_VERSION = "1.5"
IMPL = "core-python"

# Default corpus location relative to the repo root (this file lives at
# protocol-generator/python/tests/conformance/harness.py).
_REPO_ROOT = Path(__file__).resolve().parents[4]
DEFAULT_CORPUS = (
    _REPO_ROOT
    / "protocol-generator"
    / "shared"
    / "test-vectors"
    / "v0.8.0"
    / "conformance-vectors-v1.cbor"
)


def _as_str(v: Any) -> str:
    if isinstance(v, str):
        return v
    raise TypeError(f"expected text string, got {type(v).__name__}")


def _as_int(v: Any) -> int:
    if isinstance(v, bool) or not isinstance(v, int):
        raise TypeError(f"expected int, got {type(v).__name__}")
    return v


def _as_bytes(v: Any) -> bytes:
    if isinstance(v, (bytes, bytearray, ByteKey)):
        return bytes(v)
    raise TypeError(f"expected byte string, got {type(v).__name__}")


# ── Category constructions (byte-for-byte mirror of the Go oracle) ────────────
def _emit_content_hash(inp: dict) -> bytes:
    typ = _as_str(inp["type"])
    data = inp["data"]
    format_code = _as_int(inp["format_code"]) if "format_code" in inp else 0
    if format_code < 0:
        raise ValueError("content_hash format_code must be >= 0")
    encoded = _cbor.encode({"type": typ, "data": data})
    digest = hashlib.sha256(encoded).digest()
    return encode_varint(format_code) + digest


def _emit_peer_id(inp: dict) -> bytes:
    kt = _as_int(inp["key_type"])
    ht = _as_int(inp["hash_type"])
    digest = _as_bytes(inp["digest"])
    if kt < 0 or ht < 0:
        raise ValueError("peer_id codes must be >= 0")
    raw = encode_varint(kt) + encode_varint(ht) + digest
    b58 = _base58.b58encode(raw)
    # Canonical = ECF text-string encoding of the Base58 string.
    return _cbor.encode(b58)


def _emit_signature(inp: dict) -> bytes:
    seed = _as_bytes(inp["seed"])
    entity = inp["entity"]
    entity_bytes = _cbor.encode(entity)
    return sign_ed25519(seed, entity_bytes)


def _emit_envelope(inp: dict) -> bytes:
    return _cbor.encode(inp)


def _emit_class_a(inp: Any) -> bytes:
    return _cbor.encode(inp)


def emit_vector(vector_id: str, inp: Any) -> bytes:
    if vector_id.startswith("content_hash."):
        return _emit_content_hash(inp)
    if vector_id.startswith("peer_id."):
        return _emit_peer_id(inp)
    if vector_id.startswith("signature."):
        return _emit_signature(inp)
    if vector_id.startswith("envelope."):
        return _emit_envelope(inp)
    return _emit_class_a(inp)


# ── Runner ────────────────────────────────────────────────────────────────────
def run(corpus_path: Path | str = DEFAULT_CORPUS) -> dict:
    """Run the full corpus; return a structured result dict."""
    corpus_bytes = Path(corpus_path).read_bytes()
    # Step 1 (Appendix E §E.3): load the corpus through the peer's OWN decoder.
    vectors = _cbor.decode(corpus_bytes)
    if not isinstance(vectors, list):
        raise TypeError("corpus root must be a CBOR array of vector maps")

    results: list[dict] = []
    n_pass = n_fail = 0
    per_category: dict[str, dict[str, int]] = {}

    for vec in vectors:
        vid = _as_str(vec["id"])
        kind = _as_str(vec["kind"])
        canonical = _as_bytes(vec["canonical"])
        category = vid.split(".", 1)[0]
        cat = per_category.setdefault(category, {"pass": 0, "fail": 0})

        ok = False
        detail = ""
        if kind == "encode_equal":
            try:
                produced = emit_vector(vid, vec["input"])
                ok = produced == canonical
                if not ok:
                    detail = (
                        f"got {produced.hex()} want {canonical.hex()}"
                    )
            except Exception as e:  # noqa: BLE001 — report, don't crash the run
                detail = f"emit raised: {type(e).__name__}: {e}"
        elif kind == "decode_reject":
            try:
                _cbor.decode(canonical)
                detail = "decoder ACCEPTED bytes that MUST be rejected"
            except _cbor.NonCanonicalEcfError as e:
                ok = True
                detail = f"rejected: non_canonical_ecf ({e})"
            except _cbor.TruncatedError as e:
                ok = True
                detail = f"rejected: truncated ({e})"
        else:
            detail = f"unknown kind: {kind!r}"

        results.append(
            {"id": vid, "kind": kind, "pass": ok, "detail": detail}
        )
        if ok:
            n_pass += 1
            cat["pass"] += 1
        else:
            n_fail += 1
            cat["fail"] += 1

    return {
        "impl": IMPL,
        "corpus_version": CORPUS_VERSION,
        "spec_version": SPEC_VERSION,
        "total": len(vectors),
        "pass": n_pass,
        "fail": n_fail,
        "categories": per_category,
        "vectors": results,
    }


def main(argv: list[str]) -> int:
    corpus = Path(argv[1]) if len(argv) > 1 else DEFAULT_CORPUS
    report = run(corpus)
    failures = [v for v in report["vectors"] if not v["pass"]]
    print(
        f"wire-conformance ({report['impl']}, corpus {report['corpus_version']}, "
        f"spec {report['spec_version']}): "
        f"{report['pass']}/{report['total']} PASS, {report['fail']} FAIL"
    )
    cats = ", ".join(
        f"{k} {v['pass']}/{v['pass'] + v['fail']}"
        for k, v in sorted(report["categories"].items())
    )
    print(f"  categories: {cats}")
    for f in failures:
        print(f"  FAIL {f['id']} ({f['kind']}): {f['detail']}")
    print("---JSON---")
    print(json.dumps(report, indent=2))
    return 1 if report["fail"] else 0


if __name__ == "__main__":
    raise SystemExit(main(sys.argv))
